google/cloud/internal/default_completion_queue_impl.cc

```diff
@@ -64,12 +64,25 @@ class AsyncTimerFuture : public AsyncGrpcOperation {
 
   void Set(grpc::CompletionQueue& cq,
            std::chrono::system_clock::time_point deadline, void* tag) {
+    std::unique_lock<std::mutex> lk(mu_);
+    if (state_ != kIdle) return;
     deadline_ = deadline;
+    state_ = kSet;
     alarm_.Set(&cq, deadline, tag);
   }
 
   void Cancel() override {
     ScopedCallContext scope(call_context_);
+    std::unique_lock<std::mutex> lk(mu_);
+    if (state_ == kCancelled) return;
+    if (state_ == kIdle) {
```

Comment:
Note that since gRPC's Alarm cancellation is idempotent, this could be simplified a bit:

```
    std::unique_lock<std::mutex> lk(mu_);
    if (std::exchange(state_, kCancelled) == kIdle) {
      lk.unlock();
      // Release the lock before (potentially) calling application code.
      promise_.set_value(Cancelled());
      return;
    }
    alarm_.Cancel();
```

Reply:
This may be incorrect as well, actually. Please disregard.